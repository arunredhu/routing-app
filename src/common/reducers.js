import { combineReducers } from 'redux';

import routeFinder from '../routeFinder/reducers';

// Root reducers
export default combineReducers({
    routeFinder
});