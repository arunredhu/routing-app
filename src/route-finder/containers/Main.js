import React, { Component } from "react";
import { MapWrapperComponent, LocationInputComponent } from "../components/index";
import { connect } from "react-redux";
import { bindActionCreators } from "redux";
import { updateInputDetails, resetInputDetails, submitInputDetails, updateCurrentLocation } from "../actions/index";
import PropTypes from 'prop-types'

// container component that talks to redux store
class Main extends Component {

  // get current lat long
  componentDidMount() {
    this.getCurrentLocation()
  }

  render() {
    return (
      <React.Fragment>
        <LocationInputComponent submitInputDetails={this.props.submitInputDetails}
          geo={this.props.geo} updateInputDetails={this.props.updateInputDetails}
          resetInputDetails={this.props.resetInputDetails}
        />
        {
          this.props.geo.startLatLong.length > 0 && this.props.geo.dropLatLong.length > 0 &&
          <div className="margin-top-16">
            <MapWrapperComponent geo={this.props.geo} defaultZoom={10} />
          </div>
        }
      </React.Fragment>
    );
  }

  getCurrentLocation() {
    const location = window.navigator && window.navigator.geolocation
    location && location.getCurrentPosition((position) => {
      this.props.updateCurrentLocation(position.coords)
    })
  }
}

// returns parts of redux state
function mapStateToProps(routeFinder) {
  return {
    geo: routeFinder.geo
  };
}

// returns action dispactors
function mapDispatchToProps(dispatch) {
  return bindActionCreators({ updateInputDetails, resetInputDetails, submitInputDetails, updateCurrentLocation }, dispatch);
}

Main.propTypes = {
  geo: PropTypes.object,
  updateInputDetails: PropTypes.func,
  resetInputDetails: PropTypes.func,
  submitInputDetails: PropTypes.func,
  updateCurrentLocation: PropTypes.func
}

// connects component to redux
export default connect(mapStateToProps, mapDispatchToProps)(Main);