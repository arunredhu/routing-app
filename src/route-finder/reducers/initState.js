export const initialState = {
    start: "",
    drop: "",
    totalDistance: "",
    totalTime: "",
    startLatLong: "",
    dropLatLong: "",
    restLatLong: "",
    submitted: false, 
    failure: {},
    success: {}, 
    currLatLong: ""
};