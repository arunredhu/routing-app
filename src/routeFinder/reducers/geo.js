import { UPDATE_INPUT_DETAILS, RESET_INPUT_DETAILS, SUBMIT_STARTED, SUBMIT_SUCCESS, UPDATE_CURRENT_LOCATION, SUBMIT_FAILURE } from "../actions/index";
import { initialState } from "./initState"

// reducers for different action dispatchers
export default function (state = initialState, action) {
  switch (action.type) {

    case UPDATE_INPUT_DETAILS:
      const start = action.payload.type === 'start' ? action.payload.val : (state.start ? state.start : "")
      const drop = action.payload.type === 'drop' ? action.payload.val : (state.drop ? state.drop : "")

      return {
        ...state,
        start,
        drop
      }

    case RESET_INPUT_DETAILS:
      return action.payload ? { ...state, [action.payload]: "" } : { ...initialState, currLatLong: state.currLatLong }

    case SUBMIT_STARTED:
      return { ...state, submitted: true }

    case SUBMIT_FAILURE:
      return { ...initialState, failure: { message: action.payload, status: true }, success: { status: false }, start: state.start, drop: state.drop, submitted: true, currLatLong: state.currLatLong }

    case SUBMIT_SUCCESS:

      const totalDistance = action.payload.total_distance
      const totalTime = action.payload.total_time
      const startLatLong = action.payload.path[0]
      const dropLatLong = action.payload.path.pop()
      const restLatLong = action.payload.path.reduce((acc, curr, index) => {
        if (index !== 0 && index !== action.payload.path.length - 1) {
          acc.push(curr)
        }
        return acc
      })

      return {
        ...state,
        totalDistance,
        totalTime,
        startLatLong,
        dropLatLong,
        failure: { status: false },
        success: { status: true },
        restLatLong
      }

    case UPDATE_CURRENT_LOCATION:
      let currLatLong = { lat: action.payload.latitude, lng: action.payload.longitude }
      return { ...state, currLatLong }
    default:
      return state
  }
}