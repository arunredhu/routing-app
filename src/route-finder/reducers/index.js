import { combineReducers } from "redux";
import geo from "./geo";

// Route finder reducers
export default combineReducers({
  geo
});