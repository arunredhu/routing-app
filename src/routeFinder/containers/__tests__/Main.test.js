import React from 'react';
import renderer from 'react-test-renderer';
import Main from '../main';
import reducers from "../../reducers";
import thunk from 'redux-thunk';
import { createStore, applyMiddleware } from "redux";
import { Provider } from "react-redux";

const store = createStore(reducers, applyMiddleware(thunk));

describe('Main Test', () => {
  it('render correctly', () => {
    const geo = {
      startLatLong: [
        "22.372081",
        "114.107877"
      ],
      dropLatLong: [
        "22.284419",
        "114.159510"
      ]
    }
    const updateInputDetails = jest.fn()
    const resetInputDetails = jest.fn()
    const submitInputDetails = jest.fn()
    const updateCurrentLocation = jest.fn()

    const tree = renderer.create(<Provider store={store}>
      <Main updateInputDetails={updateInputDetails}
        resetInputDetails={resetInputDetails} updateCurrentLocation={updateCurrentLocation}
        submitInputDetails={submitInputDetails} geo={geo} />
    </Provider>).toJSON();

    expect(tree).toMatchSnapshot();
  });
});