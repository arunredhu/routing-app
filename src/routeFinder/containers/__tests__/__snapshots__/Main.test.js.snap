// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main Test render correctly 1`] = `
Array [
  <div>
    <label>
      <b>
        Starting Location
      </b>
    </label>
    <br />
    <input
      id="start"
      name="start"
      onChange={[Function]}
      placeholder="Enter Start Location"
      required={true}
      type="text"
      value=""
    />
    
  </div>,
  <div>
    <label>
      <b>
        Drop-off point
      </b>
    </label>
    <br />
    <input
      id="drop"
      name="drop"
      onChange={[Function]}
      placeholder="Enter Drop Location"
      required={true}
      type="text"
      value=""
    />
    
  </div>,
  "",
  <div
    className="margin-top-16"
  >
    <button
      className="btn btn-primary"
      disabled={true}
      onClick={[Function]}
      style={
        Object {
          "marginRight": 10,
        }
      }
      type="button"
    >
      Submit
    </button>
    <button
      className="btn btn-dark"
      onClick={[Function]}
      type="button"
    >
      Reset
    </button>
  </div>,
]
`;
